test/plugin/base_test.rb

```diff
@@ -0,0 +1,1318 @@
+    setup_gce_metadata_stubs
+    expected_ts = []
+    emit_index = 0
+    setup_logging_stubs do
+      [Time.at(123_456.789), Time.at(0), Time.now].each do |ts|
+        d = create_driver
+        # Test the "native" fluentd timestamp as well as our nanosecond tags.
+        d.emit({ 'message' => log_entry(emit_index) }, ts.to_f)
+        expected_ts.push(ts)
+        emit_index += 1
+        d.emit('message' => log_entry(emit_index),
+               'timeNanos' => ts.tv_sec * 1_000_000_000 + ts.tv_nsec)
+        expected_ts.push(ts)
+        emit_index += 1
+        d.emit('message' => log_entry(emit_index),
+               'timestamp' => { 'seconds' => ts.tv_sec, 'nanos' => ts.tv_nsec })
+        expected_ts.push(ts)
+        emit_index += 1
+        d.emit('message' => log_entry(emit_index),
+               'timestampSeconds' => ts.tv_sec, 'timestampNanos' => ts.tv_nsec)
+        expected_ts.push(ts)
+        emit_index += 1
+        d.run
+      end
+    end
+    verify_index = 0
+    verify_log_entries(emit_index, COMPUTE_PARAMS) do |entry|
+      assert_with_default_check(entry['metadata']['timestamp']['seconds'],
+                                expected_ts[verify_index].tv_sec, 0, entry)
+      assert_with_default_check(entry['metadata']['timestamp']['nanos'],
+                                expected_ts[verify_index].tv_nsec, 0, entry) do
+        # Fluentd v0.14 onwards supports nanosecond timestamp values.
+        # Added in 600 ns delta to avoid flaky tests introduced
+        # due to rounding error in double-precision floating-point numbers
+        # (to account for the missing 9 bits of precision ~ 512 ns).
+        # See http://wikipedia.org/wiki/Double-precision_floating-point_format
+        assert_in_delta expected_ts[verify_index].tv_nsec,
+                        entry['metadata']['timestamp']['nanos'], 600, entry
+      end
+      verify_index += 1
+    end
+  end
+
+  def test_malformed_timestamp
+    setup_gce_metadata_stubs
+    setup_logging_stubs do
+      d = create_driver
+      # if timestamp is not a hash it is passed through to the struct payload.
+      d.emit('message' => log_entry(0), 'timestamp' => 'not-a-hash')
+      d.run
+    end
+    verify_log_entries(1, COMPUTE_PARAMS, 'structPayload') do |entry|
+      fields = get_fields(entry['structPayload'])
+      assert_equal 2, fields.size, entry
+      assert_equal 'not-a-hash', get_string(fields['timestamp']), entry
+    end
+  end
+
+  # Make parse_severity public so we can test it.
+  class Fluent::GoogleCloudOutput # rubocop:disable Style/ClassAndModuleChildren
+    public :parse_severity
+  end
+
+  def test_severities
+    setup_gce_metadata_stubs
+    expected_severity = []
+    emit_index = 0
+    setup_logging_stubs do
+      d = create_driver
+      # Array of pairs of [parsed_severity, expected_severity]
+      [%w(INFO INFO), %w(warn WARNING), %w(E ERROR), %w(BLAH DEFAULT),
+       %w(105 DEBUG), ['', 'DEFAULT']].each do |sev|
```

Comment:
Let's add a TODO above the copies to consolidate them. One obvious way would be to split it into two tests, one with string severities and one with numeric severities, but I'm ok with doing it in a later cleanup.